src/DataProvider/OperationDataProvider.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\DataProvider;
+
+use ApiPlatform\Core\Exception\InvalidArgumentException;
+use ApiPlatform\Core\Identifier\Normalizer\ChainIdentifierDenormalizer;
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+
+/**
+ * @internal
+ */
+class OperationDataProvider
```

Comment:
`final`?

Reply:
I would just get rid of this class, it introduces unneeded complexity and indirection. Can we not include the trait directly in the data provider? It looks cleaner to me (and allows to keep subresources support optional).